### PR TITLE
ci: add test matrix

### PR DESCRIPTION
**Pull request by bonjourmauko**
Supersedes #1168

### CODE REVIEW

File: .github/workflows/push.yaml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        activate_command: ${{ matrix.activate_command }}
+
+  test:
+    needs: [setup]
+    strategy:
+      fail-fast: true
+      matrix:
+        os: [ubuntu-22.04, windows-2019]
+        numpy: [2.1.1, 1.24.2]
+        python: [3.11.9, 3.9.13] # Patch version must be specified to avoid any cache confusion, since the cache key depends on the full Python version. If left unspecified, different patch versions could be allocated between jobs,  and any such difference would lead to a cache not found error.
```
Comment (benoit-cty) [inline .github/workflows/push.yaml:38]:
```suggestion
        python: [3.11.9, 3.9.13]
```
I suggest to keep the command only on the first matrix of this file, for readability.

File: .github/workflows/push.yaml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0 # Fetch all the tags
+
+      - name: Set up Python
+        uses: actions/setup-python@v4
+        with:
+          python-version: 3.9.13 # Patch version must be specified to avoid any cache confusion, since the cache key depends on the full Python version. If left unspecified, different patch versions could be allocated between jobs, and any such difference would lead to a cache not found error.
```
Comment (benoit-cty) [inline .github/workflows/push.yaml:81]:
```suggestion
          python-version: 3.9.13
```

File: .github/workflows/push.yaml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        python: ${{ matrix.python }}
+        activate_command: ${{ matrix.activate_command }}
+
+  lint:
+    needs: [setup]
+    strategy:
+      fail-fast: true
+      matrix:
+        numpy: [1.24.2]
+        python: [3.11.9, 3.9.13] # Patch version must be specified to avoid any cache confusion, since the cache key depends on the full Python version. If left unspecified, different patch versions could be allocated between jobs,  and any such difference would lead to a cache not found error.
```
Comment (benoit-cty) [inline .github/workflows/push.yaml:57]:
```suggestion
        python: [3.11.9, 3.9.13]
```

File: .github/publish-git-tag.sh
Comment (benoit-cty) [inline .github/publish-git-tag.sh:1]: I don't see where the tag is set without this script ?

**Review by benoit-cty — COMMENTED** [review]
Thanks, it's nice to have shorter files for workflows.

I have some questions.

File: .github/publish-git-tag.sh
Comment (bonjourmauko) [reply .github/publish-git-tag.sh:1]: In the openfisca_tasks/publish.mk file.

**Review by benoit-cty — APPROVED** [review]

File: .github/publish-git-tag.sh
Comment (benoit-cty) [reply .github/publish-git-tag.sh:1]: Oh yes, in https://github.com/openfisca/openfisca-core/blob/32dfe8c5a4b4aaed0393fe3ba985de5facedba55/openfisca_tasks/publish.mk#L23C1-L23C38

**Pull request merged**